### PR TITLE
Push parser 4

### PR DESCRIPTION
**Pull request by kaidokert**
## Summary by Sourcery

Add a complete PushParser module for SAX‐style JSON parsing with improved error handling and expose the new API, then validate it through a large suite of conformance, stress, and edge‐case tests.

New Features:
- Introduce a no_std-compatible SAX-style PushParser for JSON with a customizable PushParserHandler trait
- Expose PushParser, PushParserHandler, and PushParseError in the public API

Enhancements:
- Extend ParseError with new variants for buffer overflow and implement From<ujson::Error]
- Integrate push‐based parsing into existing conformance suite alongside SliceParser and StreamParser

Tests:
- Add extensive conformance tests for PushParser and StreamParser covering pass/fail JSON cases, unicode escapes, deep nesting, and edge conditions
- Introduce stress tests to validate PushParser under varying buffer sizes and chunk patterns
- Add targeted tests for copy‐on‐escape optimization, escape sequence handling, and reproductions of InvalidSliceBounds issues

## Summary by CodeRabbit

* **New Features**
  * Introduced a new incremental JSON push parser with event-driven parsing and no_std support.
  * Enhanced error handling with new error variants for input buffer limits and UTF-8 issues.
  * Added public interfaces for push parsing events and handler integration.

* **Bug Fixes**
  * Clarified error messages related to buffer fullness and UTF-8 errors.

* **Tests**
  * Added extensive tests covering conformance, escape sequences, buffer boundaries, copy-on-escape behavior, and invalid inputs.
  * Introduced stress tests for buffer sizes, chunked input patterns, Unicode edge cases, and document validation.
  * Added tests simulating input buffer full scenarios and large token handling.

### CODE REVIEW

File: picojson/tests/push_parser.rs [hunk 1]
```diff
@@ -0,0 +1,856 @@
...
+        }
+
+        let mut buffer = [0u8; 256];
+        let handler = ConsecutiveUnicodeTestHandler { events: Vec::new() };
+        let mut parser = PushParser::<_, DefaultConfig>::new(handler, &mut buffer);
+
+        // Test string with mixed escapes like in pass1.json line 45
+        parser
+            .write(br#"{"key": "\/\\\"\uCAFE\uBABE\b\f\n"}"#)
+            .unwrap();
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/tests/push_parser.rs]: solved

File: picojson/tests/push_parser_copy_on_escape.rs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        first_string: None,
+        first_len: 0,
+        second_string: None,
+        second_len: 0,
+        count: 0,
+    };
+    let mut parser = PushParser::<_, DefaultConfig>::new(handler, &mut buffer);
+
+    // Test: simple string followed by escaped string
+    parser.write(br#"{"simple": "esc\n"}"#).unwrap();
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/tests/push_parser_copy_on_escape.rs]: solved

File: picojson/tests/push_parser_invalidslicebounds_repro.rs [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+fn test_reproduce_invalidslicebounds_minimal_case() {
+    // Try the most minimal case that still has unescaped content
+    let test_cases = vec![
+        (r#"{"k":""}"#, "empty string"),
+        (r#"{"k":"a"}"#, "single char"),
+        (r#"{"k":"ab"}"#, "two chars"),
+        (r#"{"k":"\n"}"#, "simple escape"),
+        (r#"{"k":"\u0041"}"#, "single unicode"),
+        (r#"{"k":"\u0123"}"#, "single unicode hex"),
+        (r#"{"k":"\u0123\u4567"}"#, "double unicode"),
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/tests/push_parser_invalidslicebounds_repro.rs]: solved

File: picojson/tests/json_checker_tests.rs [hunk 4]
```diff
@@ -85,6 +141,261 @@ mod json_checker_tests {
...
+
+                    // For unicode escapes in raw strings, they remain as literal text
+                    assert!(
+                        parsed_content.contains("\\uCAFE"),
+                        "Should contain literal \\uCAFE sequence"
+                    );
+                    assert!(
+                        parsed_content.contains("\\uBABE"),
+                        "Should contain literal \\uBABE sequence"
+                    );
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/tests/json_checker_tests.rs]: solved

File: picojson/src/push_parser.rs [hunk 5]
```diff
@@ -0,0 +1,618 @@
...
+                                                                .append_unescaped_byte(b)?;
+                                                        }
+                                                    }
+                                                }
+                                                Err(e) => return Err(e.into()),
+                                            }
+                                        }
+                                    }
+                                    Err(e) => return Err(e.into()),
+                                }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/src/push_parser.rs]: solved

File: picojson/tests/push_parser_stress_test.rs [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+        },
+        TestScenario {
+            name: "String ending with escape",
+            json: br#"["hello\\\\"]"#,
+            expected_events: vec![
+                Event::StartArray,
+                Event::String(picojson::String::Unescaped("hello\\\\")),
+                Event::EndArray,
+                Event::EndDocument,
+            ],
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/tests/push_parser_stress_test.rs:251]: solved

File: picojson/src/push_parser.rs [hunk 7]
```diff
@@ -0,0 +1,618 @@
...
+                        let initial_part = &data
+                            [(start_pos - self.position_offset)..(end_pos - self.position_offset)];
+                        for &byte in initial_part {
+                            self.stream_buffer.append_unescaped_byte(byte)?;
+                        }
+                    } else {
+                        self.using_unescaped_buffer = true;
+                        self.stream_buffer.clear_unescaped();
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/src/push_parser.rs]: solved
